### DISCUSSION TITLE
What core features do we want?

### DISCUSSION THREAD

**Discussion by jonnydgreen** [Eyrie-Tech/api-framework#3] [Ideas]
- Why would I want to use this?
- How can I trust it?
- Does this framework favour configuration over convention?
- What are the values of the framework?

**Comment by jonnydgreen** [comment]
Type-safe decorators and use the stage-3 decorators only thread

**Reply by jonnydgreen** [reply]
PoC spike:

Definition:

```typescript
export type ClassType<T> = new (...args: any[]) => T;

export type MaybeClassType<T> = ClassType<T> | T;

type MaybeClassTypes<T> = { [P in keyof T]: MaybeClassType<T[P]> };

type MapScalar<T> = T extends String ? string
  : T extends Number ? number
  : T extends Boolean ? boolean
  : T extends undefined ? undefined
  : T extends null ? null
  : T;

type MapTypes<T> = { [P in keyof T]: MapScalar<T[P]> };

export class Hello {
  public name!: string;
  public address!: string;
}

export interface PostOptions<R> {
  outputType: MaybeClassType<R>;
  isArray?: boolean;
}

const listKey = Symbol("framework.list");

function List<T>(input: MaybeClassType<T>): T[] {
  return listKey as unknown as T[];
}

const tupleKey = Symbol("framework.tuple");

function Tuple<U, T extends any[]>(
  first: MaybeClassType<U>,
  ...args: MaybeClassTypes<T>
): MapTypes<[U, ...T]> {
  return tupleKey as unknown as MapTypes<[U, ...T]>;
}

export type MaybePromise<T> = T;

function Post<R>(options: PostOptions<R>) {
  return function post<T extends (...args: any[]) => MaybePromise<R>>(
    target: T,
    context: ClassMethodDecoratorContext,
  ): void {
    const methodName = context.name;
    if (context.private) {
      throw new Error(
        `'bound' cannot decorate private properties like ${methodName as string}.`,
      );
    }
    context.addInitializer(function (this: unknown) {
      const thisArg = this as ClassType<unknown>;
      const className = thisArg.constructor.name;
      console.log({ className, methodName, this: this, target, context });
    });
  };
}
```

Usage:

```typescript
export class Controller {
  @Post({ outputType: Hello })
  public getHello(): Hello {
    return {
      name: "name",
      address: "address",
    };
  }

  @Post({ outputType: List(Hello) })
  public getHelloList(): Hello[] {
    return [{
      name: "name",
      address: "address",
    }];
  }

  @Post({ outputType: Tuple(Hello, String) })
  public getHelloTuple(): [Hello, string] {
    return [{
      name: "name",
      address: "address",
    }, "hello"];
  }
}
const c = new Controller();
c.getHello();
```

**Comment by jonnydgreen** [comment]
Possible framework API values thread:
- Be explicit
- Clear and concise API
- Convention-based
- Docs/design-first

**Comment by jonnydgreen** [comment]
Convention vs configuration thread:
- Keep to a clear convention
	 - Helps guide users to a specific way of working rather than giving several routes to the same goal

**Comment by jonnydgreen** [comment]
Technical preferences thread:
- Start with REST
- Use Deno and JSR
- Prefer native code over third-party

**Comment by aleccool213** [comment]
Purpose and "Why would I want to use this?" thread:

* best practice when maintaining a REST API is to generate [OpenAPI Spec](https://swagger.io/specification/)
  * two ways to do this, 1. write the spec first and try to match the code 2. write the code first, which gens the spec
  * similar to graphql in this respect
* there are many REST API devs who have opted for option 2 due to ergonomics
* large frameworks like nestjs and fastify support openapi gen
* nestjs uses decorators which is a great way to define APIs
* issue is that it uses TS experimental decorators which are not EMCAscript
* [TS now has EMCAscript decorators](https://blog.logrocket.com/using-modern-decorators-typescript/)
* as part of our values, [docs should be first class citizens](https://github.com/jonnydgreen/api-framework/discussions/3#discussioncomment-8803583), we should gen OpenAPI spec from v1
* devs using this framework will get first-class support for:
  * deno
  * jsr
  * ts with no gnarly options
  * decorator based conventions they are used to from other frameworks
  * performance
  * zero dependencies (maybe)

**Reply by aleccool213** [reply]
* all of the other frameworks which exist are not first class deno and also support many things, developers want to just gen openapi spec and run native emcascript that can target multiple runtimes
